File: lib/status.ts
// Read-only status report for a dump + its tracking DB.
//
// Aggregates four independent sources, each handled gracefully when absent:
//   1. <dump>/_index.json          per-type expected/written/status (from the dump)
//   2. <dump>/_enrich_report.json  per-type enriched/failed/skipped (from enrich)
//   3. on-disk per-type file counts (count *.json minus _-prefixed)
//   4. <dump>/../articles.db (or opts.db) — articles/runs tables
//
// computeStatus() never writes; the DB is opened read-only when supported.

import { DatabaseSync } from "node:sqlite";
import { exists, listTypeDirs, readJson } from "./fsutil.ts";
import { CHANGELOG_BASENAME } from "./changelog.ts";
import { loadPendingManifest } from "./staging.ts";

export interface TypeStatus {
  typeKey: string;
  diskCount: number;
  expected: number | null;
  /** articles present from the last run = written + skipped (incremental runs leave
   *  unchanged articles in place without rewriting them, so WRIT counts both). */
  written: number | null;
  status: string | null; // dump status: ok/partial/failed
  bodied: number | null; // enriched bodies reported by enrich
  errors: number | null; // enrich failures
}

export interface ErrorClass {
  klass: string;
  count: number;
}

export interface StatusReport {
  dump: string;
  db: string;
  dbPresent: boolean;
  perType: TypeStatus[];
  overall: {
    totalArticles: number | null; // rows in DB
    bodied: number | null; // has_body=1 in DB
    health: "OK" | "PARTIAL" | "STALE";
    lastRun: LastRun | null;
    newestCapturedAt: string | null;
    stalenessMs: number | null; // now - newest(capturedAt, ran_at)
    changelogPath: string | null; // <dump>/_changelog.jsonl if present
    /** per-op tally of the last run's changelog records (added/edited/deleted/body-*). */
    changelogLastRun: Record<string, number> | null;
    /** edits staged by the approval gate awaiting `f5kb approve` (_pending/). */
    pendingApproval: number;
  };
  errorClasses: ErrorClass[];
  notes: string[];
}

export interface LastRun {
  runId: string;
  ranAt: string | null;
  dumpDir: string | null;
  scanned: number | null;
  new: number | null;
  changed: number | null;
  unchanged: number | null;
  removed: number | null;
}

interface IndexType {
  typeKey?: string;
  dir?: string;
  status?: string;
  expected?: number;
  written?: number;
  /** incremental (sync) runs: unchanged articles left in place, not rewritten. */
  skipped?: number;
}
interface IndexFile {
  types?: IndexType[];
}
interface EnrichType {
  typeKey?: string;
  enriched?: number;
  failed?: number;
  skipped?: number;
}
interface EnrichFile {
  types?: EnrichType[];
}

const STALE_MS = 7 * 24 * 60 * 60 * 1000; // a dump older than a week is "stale"

// Coarse classification of a bodyError message so we can group counts.
export function classifyError(msg: string): string {
  const m = msg.toLowerCase();
  if (m.includes("404") || m.includes("not found")) return "not-found";
  if (m.includes("403") || m.includes("forbidden") || m.includes("401") || m.includes("unauthor")) {
    return "forbidden";
  }
  if (m.includes("timeout") || m.includes("timed out")) return "timeout";
  if (m.includes("429") || m.includes("rate")) return "rate-limited";
  if (m.match(/\b5\d\d\b/) || m.includes("server error")) return "server-error";
  if (
    m.includes("parse") || m.includes("no body") || m.includes("empty") || m.includes("extract")
  ) {
    return "parse/empty";
  }
  if (m.includes("network") || m.includes("fetch") || m.includes("connection")) return "network";
  return "other";
}

async function countDiskJson(typeDir: string): Promise<number> {
  let n = 0;
  try {
    for await (const e of Deno.readDir(typeDir)) {
      if (e.isFile && e.name.endsWith(".json") && !e.name.startsWith("_")) n++;
    }
  } catch {
    // missing dir -> 0
  }
  return n;
}

function openReadOnly(path: string): DatabaseSync {
  // node:sqlite supports a readOnly open option; fall back to a default open if
  // the running Deno doesn't recognize it.
  try {
    return new DatabaseSync(path, { readOnly: true } as unknown as undefined);
  } catch {
    return new DatabaseSync(path);
  }
}

export async function computeStatus(
  opts: { dump: string; db?: string },
): Promise<StatusReport> {
  const dump = opts.dump.replace(/\/+$/, "");
  const dbPath = opts.db ?? `${dump}/../articles.db`;
  const notes: string[] = [];

  // --- _index.json ---
  const indexPath = `${dump}/_index.json`;
  const indexByType = new Map<string, IndexType>();
  if (await exists(indexPath)) {
    try {
      const idx = await readJson<IndexFile>(indexPath);
      for (const t of idx.types ?? []) {
        const key = t.typeKey ?? t.dir;
        if (key) indexByType.set(key, t);
      }
    } catch (e) {
      notes.push(`could not parse _index.json: ${(e as Error).message}`);
    }
  } else {
    notes.push("_index.json missing");
  }

  // --- _enrich_report.json ---
  const enrichPath = `${dump}/_enrich_report.json`;
  const enrichByType = new Map<string, EnrichType>();
  if (await exists(enrichPath)) {
    try {
      const er = await readJson<EnrichFile>(enrichPath);
      for (const t of er.types ?? []) {
        if (t.typeKey) enrichByType.set(t.typeKey, t);
      }
    } catch (e) {
      notes.push(`could not parse _enrich_report.json: ${(e as Error).message}`);
    }
  } else {
    notes.push("_enrich_report.json missing");
  }

  // --- on-disk per-type counts ---
  let typeKeys: string[] = [];
  try {
    typeKeys = await listTypeDirs(dump);
  } catch {
    notes.push(`dump dir ${dump} not readable`);
  }
  // Include any type known from _index even if its dir is absent on disk.
  for (const k of indexByType.keys()) if (!typeKeys.includes(k)) typeKeys.push(k);
  typeKeys.sort();

  const perType: TypeStatus[] = [];
  for (const typeKey of typeKeys) {
    const diskCount = await countDiskJson(`${dump}/${typeKey}`);
    const idx = indexByType.get(typeKey);
    const enr = enrichByType.get(typeKey);
    // WRIT = written + skipped: an incremental sync rewrites only changed articles
    // (skipping unchanged ones in place), so "written" alone understates how many
    // articles the run actually accounts for. Adding skipped makes WRIT line up with
    // DISK/EXP. For a non-incremental dump skipped is absent (0), so this is a no-op.
    const presentFromIndex = idx ? (idx.written ?? 0) + (idx.skipped ?? 0) : null;
    perType.push({
      typeKey,
      diskCount,
      expected: idx?.expected ?? null,
      written: presentFromIndex,
      status: idx?.status ?? null,
      bodied: enr?.enriched ?? null,
      errors: enr?.failed ?? null,
    });
  }

  // --- DB (read-only) ---
  let totalArticles: number | null = null;
  let bodied: number | null = null;
  let lastRun: LastRun | null = null;
  let newestCapturedAt: string | null = null;
  const errorClasses: ErrorClass[] = [];
  // Authoritative per-document-type bodied/error counts from the DB (the enrich
  // report only covers the last run's types, so it is an unreliable per-type source).
  const dbByType = new Map<string, { bodied: number; errors: number }>();
  const dbPresent = await exists(dbPath);
  if (dbPresent) {
    let db: DatabaseSync | null = null;
    try {
      db = openReadOnly(dbPath);
      totalArticles =
        (db.prepare("SELECT COUNT(*) AS c FROM articles").get() as { c: number } | undefined)?.c ??
          null;
      bodied = (db.prepare("SELECT COUNT(*) AS c FROM articles WHERE has_body=1").get() as
        | { c: number }
        | undefined)?.c ?? null;

      const errRows = db.prepare(
        "SELECT body_error FROM articles WHERE body_error IS NOT NULL AND body_error != ''",
      ).all() as Array<{ body_error: string }>;
      const tally = new Map<string, number>();
      for (const r of errRows) {
        const k = classifyError(String(r.body_error));
        tally.set(k, (tally.get(k) ?? 0) + 1);
      }
      for (const [klass, count] of [...tally.entries()].sort((a, b) => b[1] - a[1])) {
        errorClasses.push({ klass, count });
      }

      const run = db.prepare(
        "SELECT * FROM runs ORDER BY ran_at DESC LIMIT 1",
      ).get() as {
        run_id: string;
        ran_at?: string;
        dump_dir?: string;
        scanned?: number;
        new?: number;
        changed?: number;
        unchanged?: number;
        removed?: number;
      } | undefined;
      if (run) {
        lastRun = {
          runId: run.run_id,
          ranAt: run.ran_at ?? null,
          dumpDir: run.dump_dir ?? null,
          scanned: run.scanned ?? null,
          new: run.new ?? null,
          changed: run.changed ?? null,
          unchanged: run.unchanged ?? null,
          removed: run.removed ?? null,
        };
      }
      const cap = db.prepare(
        "SELECT MAX(captured_at) AS m FROM articles",
      ).get() as { m: string } | undefined;
      newestCapturedAt = cap?.m ?? null;

      const grpRows = db.prepare(
        "SELECT document_type AS dt, " +
          "SUM(CASE WHEN has_body=1 THEN 1 ELSE 0 END) AS bodied, " +
          "SUM(CASE WHEN body_error IS NOT NULL AND body_error != '' THEN 1 ELSE 0 END) AS errors " +
          "FROM articles GROUP BY document_type",
      ).all() as Array<{ dt: string; bodied: number; errors: number }>;
      for (const r of grpRows) {
        dbByType.set(String(r.dt), { bodied: Number(r.bodied), errors: Number(r.errors) });
      }
    } catch (e) {
      notes.push(`could not read DB ${dbPath}: ${(e as Error).message}`);
    } finally {
      try {
        db?.close();
      } catch { /* ignore */ }
    }
  } else {
    notes.push(`DB ${dbPath} missing`);
  }

  // Prefer DB-derived per-type bodied/error counts (complete); the dir name maps to
  // the document_type by turning underscores back into spaces.
  for (const t of perType) {
    const d = dbByType.get(t.typeKey.replaceAll("_", " "));
    if (d) {
      t.bodied = d.bodied;
      t.errors = d.errors;
    }
  }

  // --- changelog (optional) ---
  let changelogPath: string | null = null;
  let changelogLastRun: Record<string, number> | null = null;
  const clPath = `${dump}/${CHANGELOG_BASENAME}`;
  if (await exists(clPath)) {
    changelogPath = clPath;
    if (lastRun) {
      try {
        const text = await Deno.readTextFile(clPath);
        const tally: Record<string, number> = {};
        for (const line of text.split("\n")) {
          if (!line.trim()) continue;
          let rec: { runId?: string; op?: string };
          try {
            rec = JSON.parse(line);
          } catch {
            continue; // skip a malformed line rather than fail the whole report
          }
          if (rec.runId === lastRun.runId && rec.op) tally[rec.op] = (tally[rec.op] ?? 0) + 1;
        }
        changelogLastRun = tally;
      } catch {
        notes.push(`could not read changelog ${clPath}`);
      }
    }
  }

  // --- pending approval (_pending/_manifest.json) ---
  let pendingApproval = 0;
  try {
    pendingApproval = (await loadPendingManifest(dump)).entries.length;
  } catch {
    // no/unreadable manifest -> 0
  }

  // --- staleness + health ---
  const now = Date.now();
  const stamps: number[] = [];
  const capMs = newestCapturedAt ? Date.parse(newestCapturedAt) : NaN;
  if (!Number.isNaN(capMs)) stamps.push(capMs);
  const ranMs = lastRun?.ranAt ? Date.parse(lastRun.ranAt) : NaN;
  if (!Number.isNaN(ranMs)) stamps.push(ranMs);
  const newest = stamps.length ? Math.max(...stamps) : null;
  const stalenessMs = newest != null ? now - newest : null;

  // Health reflects DUMP completeness via the dump's OWN per-type status in
  // _index.json (ok/partial/failed) — which already accounts for --all vs --days vs
  // --limit (written<expected is normal under --days/--limit, so we must NOT
  // re-derive it here). Per-article bodyErrors (404s, moved pages, empty stubs) are
  // expected enrichment-coverage detail shown in the ERR column + error-class
  // breakdown — they do NOT degrade health on their own.
  const anyPartial = perType.some((t) => t.status != null && t.status !== "ok");
  let health: StatusReport["overall"]["health"];
  if (stalenessMs != null && stalenessMs > STALE_MS) health = "STALE";
  else if (anyPartial || !dbPresent) health = "PARTIAL";
  else health = "OK";

  return {
    dump,
    db: dbPath,
    dbPresent,
    perType,
    overall: {
      totalArticles,
      bodied,
      health,
      lastRun,
      newestCapturedAt,
      stalenessMs,
      changelogPath,
      changelogLastRun,
      pendingApproval,
    },
    errorClasses,
    notes,
  };
}

// ---------------------------------------------------------------------------
// Rendering
// ---------------------------------------------------------------------------
function fmtAge(ms: number | null): string {
  if (ms == null) return "?";
  const sec = ms / 1000;
  if (sec < 60) return `${sec.toFixed(0)}s ago`;
  const min = sec / 60;
  if (min < 60) return `${min.toFixed(0)}m ago`;
  const hr = min / 60;
  if (hr < 24) return `${hr.toFixed(1)}h ago`;
  return `${(hr / 24).toFixed(1)}d ago`;
}

function pad(s: string, w: number): string {
  return s.length >= w ? s : s + " ".repeat(w - s.length);
}
function padL(s: string, w: number): string {
  return s.length >= w ? s : " ".repeat(w - s.length) + s;
}

export function renderStatus(report: StatusReport): string {
  const lines: string[] = [];
  const o = report.overall;
  lines.push(`Status: ${o.health}   dump=${report.dump}`);
  lines.push(
    `  DB: ${report.dbPresent ? report.db : "(missing)"}  ` +
      `articles=${o.totalArticles ?? "?"}  bodied=${o.bodied ?? "?"}`,
  );
  if (o.lastRun) {
    const r = o.lastRun;
    lines.push(
      `  last run ${r.runId} (${fmtAge(o.stalenessMs)}): ` +
        `scanned=${r.scanned ?? "?"} new=${r.new ?? "?"} changed=${r.changed ?? "?"} ` +
        `unchanged=${r.unchanged ?? "?"} removed=${r.removed ?? "?"}`,
    );
  } else {
    lines.push(`  last run: (none recorded)`);
  }
  if (o.newestCapturedAt) {
    lines.push(`  newest capturedAt: ${o.newestCapturedAt} (${fmtAge(o.stalenessMs)})`);
  }
  if (o.changelogPath) {
    const cl = o.changelogLastRun;
    const summary = cl && Object.keys(cl).length
      ? Object.entries(cl).map(([op, n]) => `${op}=${n}`).join(" ")
      : "(no records for last run)";
    lines.push(`  changelog: ${o.changelogPath}`);
    lines.push(`    last run: ${summary}`);
  }
  if (o.pendingApproval > 0) {
    lines.push(
      `  PENDING APPROVAL: ${o.pendingApproval} staged edit(s) in ${report.dump}/_pending/ ` +
        `— review then \`f5kb approve\``,
    );
  }

  // Per-type table.
  const head = `  ${pad("TYPE", 26)} ${padL("DISK", 8)} ${padL("EXP", 8)} ${padL("WRIT", 8)} ${
    pad("STATUS", 9)
  } ${padL("BODIED", 8)} ${padL("ERR", 6)}`;
  lines.push("");
  lines.push(head);
  lines.push("  " + "-".repeat(head.length - 2));
  for (const t of report.perType) {
    lines.push(
      `  ${pad(t.typeKey, 26)} ${padL(String(t.diskCount), 8)} ` +
        `${padL(t.expected != null ? String(t.expected) : "-", 8)} ` +
        `${padL(t.written != null ? String(t.written) : "-", 8)} ` +
        `${pad(t.status ?? "-", 9)} ` +
        `${padL(t.bodied != null ? String(t.bodied) : "-", 8)} ` +
        `${padL(t.errors != null ? String(t.errors) : "-", 6)}`,
    );
  }

  if (report.errorClasses.length) {
    lines.push("");
    lines.push("  body errors by class:");
    for (const ec of report.errorClasses) lines.push(`    ${pad(ec.klass, 16)} ${ec.count}`);
  }
  if (report.notes.length) {
    lines.push("");
    lines.push("  notes:");
    for (const n of report.notes) lines.push(`    - ${n}`);
  }
  return lines.join("\n");
}

File: f5kb/lib/status.py
"""Read-only status report for a dump + its tracking DB."""

from __future__ import annotations

import re
import sqlite3
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from f5kb.lib.fsutil import list_type_dirs, path_exists, read_json
from f5kb.lib.changelog import CHANGELOG_BASENAME
from f5kb.lib.staging import load_pending_manifest

STALE_MS = 7 * 24 * 60 * 60 * 1000  # 1 week


@dataclass
class TypeStatus:
    type_key: str
    disk_count: int
    expected: int | None
    written: int | None
    status: str | None
    bodied: int | None
    errors: int | None


@dataclass
class ErrorClass:
    klass: str
    count: int


@dataclass
class LastRun:
    run_id: str
    ran_at: str | None
    dump_dir: str | None
    scanned: int | None
    new: int | None
    changed: int | None
    unchanged: int | None
    removed: int | None


@dataclass
class StatusReport:
    dump: str
    db: str
    db_present: bool
    per_type: list[TypeStatus]
    overall: dict  # totalArticles, bodied, health, lastRun, ...
    error_classes: list[ErrorClass]
    notes: list[str]


def classify_error(msg: str) -> str:
    m = msg.lower()
    if "404" in m or "not found" in m:
        return "not-found"
    if any(x in m for x in ("403", "forbidden", "401", "unauthor")):
        return "forbidden"
    if "timeout" in m or "timed out" in m:
        return "timeout"
    if "429" in m or "rate" in m:
        return "rate-limited"
    if re.search(r"\b5\d\d\b", m) or "server error" in m:
        return "server-error"
    if any(x in m for x in ("parse", "no body", "empty", "extract")):
        return "parse/empty"
    if any(x in m for x in ("network", "fetch", "connection")):
        return "network"
    return "other"


def _count_disk_json(type_dir: str) -> int:
    n = 0
    try:
        for p in Path(type_dir).iterdir():
            if p.is_file() and p.suffix == ".json" and not p.name.startswith("_"):
                n += 1
    except Exception:
        pass
    return n


def compute_status(dump: str, db: str | None = None) -> StatusReport:
    dump = dump.rstrip("/")
    db_path = db or str(Path(dump).parent / "articles.db")
    notes: list[str] = []

    # _index.json
    index_by_type: dict[str, dict] = {}
    index_path = str(Path(dump) / "_index.json")
    if path_exists(index_path):
        try:
            idx = read_json(index_path)
            for t in idx.get("types") or []:
                key = t.get("typeKey") or t.get("dir")
                if key:
                    index_by_type[key] = t
        except Exception as e:
            notes.append(f"could not parse _index.json: {e}")
    else:
        notes.append("_index.json missing")

    # _enrich_report.json
    enrich_by_type: dict[str, dict] = {}
    enrich_path = str(Path(dump) / "_enrich_report.json")
    if path_exists(enrich_path):
        try:
            er = read_json(enrich_path)
            for t in er.get("types") or []:
                if t.get("typeKey"):
                    enrich_by_type[t["typeKey"]] = t
        except Exception as e:
            notes.append(f"could not parse _enrich_report.json: {e}")
    else:
        notes.append("_enrich_report.json missing")

    # disk type dirs
    type_keys: list[str] = []
    try:
        type_keys = list_type_dirs(dump)
    except Exception:
        notes.append(f"dump dir {dump} not readable")
    for k in index_by_type:
        if k not in type_keys:
            type_keys.append(k)
    type_keys.sort()

    per_type: list[TypeStatus] = []
    for type_key in type_keys:
        disk_count = _count_disk_json(str(Path(dump) / type_key))
        idx = index_by_type.get(type_key)
        enr = enrich_by_type.get(type_key)
        present_from_index = ((idx.get("written") or 0) + (idx.get("skipped") or 0)) if idx else None
        per_type.append(TypeStatus(
            type_key=type_key,
            disk_count=disk_count,
            expected=idx.get("expected") if idx else None,
            written=present_from_index,
            status=idx.get("status") if idx else None,
            bodied=enr.get("enriched") if enr else None,
            errors=enr.get("failed") if enr else None,
        ))

    # DB
    total_articles: int | None = None
    bodied: int | None = None
    last_run: LastRun | None = None
    newest_captured_at: str | None = None
    error_classes: list[ErrorClass] = []
    db_by_type: dict[str, dict] = {}
    db_present = path_exists(db_path)
    if db_present:
        try:
            conn = sqlite3.connect(db_path)
            conn.row_factory = sqlite3.Row
            total_articles = conn.execute("SELECT COUNT(*) AS c FROM articles").fetchone()["c"]
            bodied = conn.execute("SELECT COUNT(*) AS c FROM articles WHERE has_body=1").fetchone()["c"]

            err_rows = conn.execute(
                "SELECT body_error FROM articles WHERE body_error IS NOT NULL AND body_error != ''"
            ).fetchall()
            tally: dict[str, int] = {}
            for r in err_rows:
                k = classify_error(str(r["body_error"]))
                tally[k] = tally.get(k, 0) + 1
            error_classes = [ErrorClass(klass=k, count=c) for k, c in sorted(tally.items(), key=lambda x: -x[1])]

            run_row = conn.execute("SELECT * FROM runs ORDER BY ran_at DESC LIMIT 1").fetchone()
            if run_row:
                last_run = LastRun(
                    run_id=run_row["run_id"],
                    ran_at=run_row["ran_at"],
                    dump_dir=run_row["dump_dir"],
                    scanned=run_row["scanned"],
                    new=run_row["new"],
                    changed=run_row["changed"],
                    unchanged=run_row["unchanged"],
                    removed=run_row["removed"],
                )

            cap_row = conn.execute("SELECT MAX(captured_at) AS m FROM articles").fetchone()
            newest_captured_at = cap_row["m"] if cap_row else None

            grp_rows = conn.execute(
                "SELECT document_type AS dt, "
                "SUM(CASE WHEN has_body=1 THEN 1 ELSE 0 END) AS bodied, "
                "SUM(CASE WHEN body_error IS NOT NULL AND body_error != '' THEN 1 ELSE 0 END) AS errors "
                "FROM articles GROUP BY document_type"
            ).fetchall()
            for r in grp_rows:
                db_by_type[str(r["dt"])] = {"bodied": int(r["bodied"]), "errors": int(r["errors"])}

            conn.close()
        except Exception as e:
            notes.append(f"could not read DB {db_path}: {e}")
    else:
        notes.append(f"DB {db_path} missing")

    for t in per_type:
        d = db_by_type.get(t.type_key.replace("_", " "))
        if d:
            t.bodied = d["bodied"]
            t.errors = d["errors"]

    # changelog
    changelog_path_val: str | None = None
    changelog_last_run: dict[str, int] | None = None
    cl_path = str(Path(dump) / CHANGELOG_BASENAME)
    if path_exists(cl_path):
        changelog_path_val = cl_path
        if last_run:
            try:
                tally2: dict[str, int] = {}
                with open(cl_path, encoding="utf-8") as f:
                    for line in f:
                        line = line.strip()
                        if not line:
                            continue
                        try:
                            rec = __import__("json").loads(line)
                        except Exception:
                            continue
                        if rec.get("runId") == last_run.run_id and rec.get("op"):
                            op = rec["op"]
                            tally2[op] = tally2.get(op, 0) + 1
                changelog_last_run = tally2
            except Exception:
                notes.append(f"could not read changelog {cl_path}")

    # pending approval
    pending_approval = 0
    try:
        manifest = load_pending_manifest(dump)
        pending_approval = len(manifest.get("entries") or [])
    except Exception:
        pass

    # staleness + health
    now_ms = int(time.time() * 1000)
    stamps: list[int] = []
    if newest_captured_at:
        try:
            import datetime
            dt = datetime.datetime.fromisoformat(newest_captured_at.replace("Z", "+00:00"))
            stamps.append(int(dt.timestamp() * 1000))
        except Exception:
            pass
    if last_run and last_run.ran_at:
        try:
            import datetime
            dt = datetime.datetime.fromisoformat(last_run.ran_at.replace("Z", "+00:00"))
            stamps.append(int(dt.timestamp() * 1000))
        except Exception:
            pass
    newest = max(stamps) if stamps else None
    staleness_ms = (now_ms - newest) if newest is not None else None

    any_partial = any(t.status not in (None, "ok") for t in per_type)
    if staleness_ms is not None and staleness_ms > STALE_MS:
        health = "STALE"
    elif any_partial or not db_present:
        health = "PARTIAL"
    else:
        health = "OK"

    return StatusReport(
        dump=dump,
        db=db_path,
        db_present=db_present,
        per_type=per_type,
        overall={
            "totalArticles": total_articles,
            "bodied": bodied,
            "health": health,
            "lastRun": last_run,
            "newestCapturedAt": newest_captured_at,
            "stalenessMs": staleness_ms,
            "changelogPath": changelog_path_val,
            "changelogLastRun": changelog_last_run,
            "pendingApproval": pending_approval,
        },
        error_classes=error_classes,
        notes=notes,
    )


def _fmt_age(ms: int | None) -> str:
    if ms is None:
        return "?"
    sec = ms / 1000
    if sec < 60:
        return f"{sec:.0f}s ago"
    min_ = sec / 60
    if min_ < 60:
        return f"{min_:.0f}m ago"
    hr = min_ / 60
    if hr < 24:
        return f"{hr:.1f}h ago"
    return f"{hr / 24:.1f}d ago"


def _pad(s: str, w: int) -> str:
    return s if len(s) >= w else s + " " * (w - len(s))


def _padl(s: str, w: int) -> str:
    return s if len(s) >= w else " " * (w - len(s)) + s


def render_status(report: StatusReport) -> str:
    lines: list[str] = []
    o = report.overall
    lines.append(f"Status: {o['health']}   dump={report.dump}")
    lines.append(
        f"  DB: {report.db if report.db_present else '(missing)'}  "
        f"articles={o['totalArticles'] or '?'}  bodied={o['bodied'] or '?'}"
    )
    lr = o.get("lastRun")
    if lr:
        lines.append(
            f"  last run {lr.run_id} ({_fmt_age(o.get('stalenessMs'))}): "
            f"scanned={lr.scanned or '?'} new={lr.new or '?'} changed={lr.changed or '?'} "
            f"unchanged={lr.unchanged or '?'} removed={lr.removed or '?'}"
        )
    else:
        lines.append("  last run: (none recorded)")
    if o.get("newestCapturedAt"):
        lines.append(f"  newest capturedAt: {o['newestCapturedAt']} ({_fmt_age(o.get('stalenessMs'))})")
    if o.get("changelogPath"):
        cl = o.get("changelogLastRun")
        summary = (
            " ".join(f"{k}={v}" for k, v in cl.items())
            if cl and cl
            else "(no records for last run)"
        )
        lines.append(f"  changelog: {o['changelogPath']}")
        lines.append(f"    last run: {summary}")
    if o.get("pendingApproval", 0) > 0:
        lines.append(
            f"  PENDING APPROVAL: {o['pendingApproval']} staged edit(s) in {report.dump}/_pending/ "
            f"— review then `f5kb approve`"
        )

    head = (
        f"  {_pad('TYPE', 26)} {_padl('DISK', 8)} {_padl('EXP', 8)} "
        f"{_padl('WRIT', 8)} {_pad('STATUS', 9)} {_padl('BODIED', 8)} {_padl('ERR', 6)}"
    )
    lines.append("")
    lines.append(head)
    lines.append("  " + "-" * (len(head) - 2))
    for t in report.per_type:
        lines.append(
            f"  {_pad(t.type_key, 26)} {_padl(str(t.disk_count), 8)} "
            f"{_padl(str(t.expected) if t.expected is not None else '-', 8)} "
            f"{_padl(str(t.written) if t.written is not None else '-', 8)} "
            f"{_pad(t.status or '-', 9)} "
            f"{_padl(str(t.bodied) if t.bodied is not None else '-', 8)} "
            f"{_padl(str(t.errors) if t.errors is not None else '-', 6)}"
        )

    if report.error_classes:
        lines.append("")
        lines.append("  body errors by class:")
        for ec in report.error_classes:
            lines.append(f"    {_pad(ec.klass, 16)} {ec.count}")
    if report.notes:
        lines.append("")
        lines.append("  notes:")
        for n in report.notes:
            lines.append(f"    - {n}")
    return "\n".join(lines)
